system/lib/libc/musl/src/thread/__timedwait.c

```diff
@@ -0,0 +1,82 @@
+#include <pthread.h>
+#include <time.h>
+#include <errno.h>
+#ifdef __EMSCRIPTEN__
+#include <math.h>
+#include <emscripten/threading.h>
+#include <emscripten/emscripten.h>
+#include "pthread_impl.h"
+#else
+#include "futex.h"
+#endif
+#include "syscall.h"
+
+#ifdef __EMSCRIPTEN__
+double _pthread_nsecs_until(const struct timespec *restrict at);
+int _pthread_isduecanceled(struct pthread *pthread_ptr);
+#endif
+
+static int do_wait(volatile int *addr, int val,
+	clockid_t clk, const struct timespec *at, int priv)
+{
+	int r;
+	struct timespec to, *top=0;
+
+	if (at) {
+		if (at->tv_nsec >= 1000000000UL) return EINVAL;
+		if (clock_gettime(clk, &to)) return EINVAL;
+		to.tv_sec = at->tv_sec - to.tv_sec;
+		if ((to.tv_nsec = at->tv_nsec - to.tv_nsec) < 0) {
+			to.tv_sec--;
+			to.tv_nsec += 1000000000;
+		}
+		if (to.tv_sec < 0) return ETIMEDOUT;
+		top = &to;
+	}
+
+#ifdef __EMSCRIPTEN__
+	if (1 || pthread_self()->cancelasync == PTHREAD_CANCEL_ASYNCHRONOUS) {
+		do {
+			if (_pthread_isduecanceled(pthread_self())) {
+				// Emscripten-specific return value: The wait was canceled by user calling
```

Comment:
Native implementations are able to pre-empt asynchronous cancellation with signals-like mechanism, but we don't have that ability in JS, so we must poll for such cancellation points ourselves. That requires that we propagate a custom exit value out from this function to distinguish exiting due to being timed out, or being canceled, so that the caller of this wait is able to do proper cleanup and return the correct return value.